### PR TITLE
Move RemoveEmptyFinalizersAnalyzer to syntax analyzer

### PR DESCRIPTION
**Pull request by 333fred**
Because of the Higher vs Lower issue in IOperation right now, and since we want to ship this analyzer as part of our default set, we need to move RemoveEmptyFinalizersAnalyzer back to Symbol+Syntax for now. Tagging @dotnet/roslyn-analysis @mavasani for review.

### CODE REVIEW

File: src/Microsoft.CodeQuality.Analyzers/Core/QualityGuidelines/AbstractRemoveEmptyFinalizers.cs [hunk 1]
```diff
@@ -35,72 +33,26 @@ public override void Initialize(AnalysisContext analysisContext)
             analysisContext.EnableConcurrentExecution();
             analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
 
-            analysisContext.RegisterCompilationStartAction(compilationContext =>
+            analysisContext.RegisterSymbolAction(methodContext =>
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/QualityGuidelines/AbstractRemoveEmptyFinalizers.cs]: Why not a code block analyzer?

Comment (333fred) [inline src/Microsoft.CodeQuality.Analyzers/Core/QualityGuidelines/AbstractRemoveEmptyFinalizers.cs]: Because symbol seemed to me to be more natural and correct here. We're not looking at every code block, we're looking at every symbol declaration to see the body of the symbol. If we went with code block, then the first thing we'd have to do is get the symbol for that code block to see if it's an analyzer.

Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/QualityGuidelines/AbstractRemoveEmptyFinalizers.cs]: > first thing we'd have to do is get the symbol for that code block to see if it's a finalizer

Symbol for the containing code block should already be present on the context.

Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/QualityGuidelines/AbstractRemoveEmptyFinalizers.cs]: http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs,836

Comment (333fred) [inline src/Microsoft.CodeQuality.Analyzers/Core/QualityGuidelines/AbstractRemoveEmptyFinalizers.cs]: Sure, but code block could potentially get called back multiple times for a valid finalizer, right?
```csharp
class C1
{
    ~C1() 
    {
        if (true) 
        {
            DisposeOfSomething();
        }
    }
}
```

File: src/Microsoft.CodeQuality.Analyzers/CSharp/QualityGuidelines/CSharpRemoveEmptyFinalizersAnalyzer.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+using Microsoft.CodeQuality.Analyzers.QualityGuidelines;
+
+namespace Microsoft.CodeQuality.CSharp.Analyzers.QualityGuidelines
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal class CSharpRemoveEmptyFinalizersAnalyzer : AbstractRemoveEmptyFinalizersAnalyzer
+    {
+        protected override bool IsEmptyFinalizer(SyntaxNode methodBody, SymbolAnalysisContext analysisContext)
+        {
+            var destructorDeclaration = (DestructorDeclarationSyntax)methodBody;
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/CSharp/QualityGuidelines/CSharpRemoveEmptyFinalizersAnalyzer.cs]: We can share large amount of code between C# and VB implementations. If we are planning to revert this to IOperation analyzer soon, then I am fine with code duplication. @jinujoseph are we going to revert this change once underlying IOperation issue is fixed?

Comment (333fred) [inline src/Microsoft.CodeQuality.Analyzers/CSharp/QualityGuidelines/CSharpRemoveEmptyFinalizersAnalyzer.cs]: Not sure how we'd share the parts that I split up, pretty sure I shared as much as could. At this point, we need to get the method body, which is syntax-level.

Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/CSharp/QualityGuidelines/CSharpRemoveEmptyFinalizersAnalyzer.cs]: We can share all the code dealing with semantics. Regardless, I presume we are going to switch this back to IOperation analyzer soon, so I am fine with the current approach.

File: src/Microsoft.CodeQuality.Analyzers/Core/QualityGuidelines/AbstractRemoveEmptyFinalizers.cs (hunk 1 above)
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/QualityGuidelines/AbstractRemoveEmptyFinalizers.cs]: Each code block analyzer is called once for the whole method body. It doesn't get called back for every block statement.

Comment (333fred) [inline src/Microsoft.CodeQuality.Analyzers/Core/QualityGuidelines/AbstractRemoveEmptyFinalizers.cs]: Unless I'm misunderstanding how code block analyzers work

File: src/Microsoft.CodeQuality.Analyzers/CSharp/QualityGuidelines/CSharpRemoveEmptyFinalizersAnalyzer.cs (hunk 2 above)
Comment (jinujoseph) [inline src/Microsoft.CodeQuality.Analyzers/CSharp/QualityGuidelines/CSharpRemoveEmptyFinalizersAnalyzer.cs]: yes , we should revert this back to ioperation analyzer once the underline issue is fixed

File: src/Microsoft.CodeQuality.Analyzers/Core/QualityGuidelines/AbstractRemoveEmptyFinalizers.cs [hunk 3]
```diff
@@ -35,72 +33,26 @@ public override void Initialize(AnalysisContext analysisContext)
...
-            });
-        }
-
-        private static bool IsEmptyFinalizer(ImmutableArray<IOperation> operationBlocks, ISymbol conditionalAttributeSymbol)
-        {
-            if (operationBlocks != null && operationBlocks.Length == 1)
-            {
-                var block = operationBlocks[0] as IBlockStatement;
-                if (block == null)
+                var methodSymbol = (IMethodSymbol)methodContext.OwningSymbol;
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/QualityGuidelines/AbstractRemoveEmptyFinalizers.cs]: This may not always be a method symbol. For example, this will also be called for a field initializer.

Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/QualityGuidelines/AbstractRemoveEmptyFinalizers.cs]: Essentially, this gets called back for every top level executable code block.
